feat: --find-renames for discovering renames based on content

fwiesel opened the pull request:
If we rename objects, they will appear as
add and remove actions, which makes it difficult
to assess the semantic delta, which may only be
a simple renaming.

The current implementation creates potentially
a diff between every add/remove item, which
means O(n^2) runtime. So it should only be
enabled if such changes are rare

fwiesel commented:
I do not see a way around the O(n^2) runtime, except with some heuristics. But I think in most use-cases that is acceptable.
What I would like to change before considering it ready for review is:
- max-content-delta: That should be more properly explained, and possibly replaced with something more intuitive.
- Missing Test: At least a simple test case for the new feature should be implemented

mumoshu commented on diff/diff_test.go:
```diff
@@ -195,29 +195,61 @@ metadata:
+  metadata:
+-   name: nginx
++   name: nginx-renamed
+
```
Would you mind adding one more test for the case that there's change in `metadata.name` plus there're one update, one deletion, and one insertion under `spec`?
I think that would make this feature extremely hard to get regression in the future!

fwiesel replied:
Sure, I'll add them.

mumoshu commented on cmd/rollback.go:
```diff
@@ -81,6 +82,7 @@ func rollbackCmd() *cobra.Command {
 	rollbackCmd.Flags().BoolVar(&diff.detailedExitCode, "detailed-exitcode", false, "return a non-zero exit code when there are changes")
 	rollbackCmd.Flags().StringArrayVar(&diff.suppressedKinds, "suppress", []string{}, "allows suppression of the values listed in the diff output")
 	rollbackCmd.Flags().IntVarP(&diff.outputContext, "context", "C", -1, "output NUM lines of context around changes")
+	rollbackCmd.Flags().Float32VarP(&diff.maxChangeFraction, "max-content-change-fraction", "D", 0, "maximum fraction of changed content as lines added + removed compared to total lines in a diff for considering it a moved resource")
```
```suggestion
	rollbackCmd.Flags().Float32VarP(&diff.findRenames, "find-renames", "D", 0, "Enable rename detection if set to any value greater than 0. If specified, the value denotes the maximum fraction of changed content as lines added + removed compared to total lines in a diff for considering it a rename")
```

How about using `find-renames` like the similar feature exists in `git diff`?

```
       -M[<n>], --find-renames[=<n>]
           Detect renames. If n is specified, it is a threshold on the similarity index (i.e. amount of addition/deletions compared to the file's size). For example, -M90% means Git should consider a delete/add pair
           to be a rename if more than 90% of the file hasn't changed. Without a % sign, the number is to be read as a fraction, with a decimal point before it. I.e., -M5 becomes 0.5, and is thus the same as -M50%.
           Similarly, -M05 is the same as -M5%. To limit detection to exact renames, use -M100%. The default similarity index is 50%.
```

fwiesel replied:
It is indeed inspired by `find-renames`... 

 I was doubting if the expectation of a `find-renames` would be to only discover changes of `metadata.name` and possibly `metadata.namespace`, but not the kind. 
My problem though is, that the proposal also detects changes of the `Kind` (e.g. `Deployment` -> `StatefulSet`).

But since your intuition is that it is the same feature, I may be overthinking things here.

In that light, what is your opinion? If you still say it is more intuitive to use the same parameters, I will change it. I wasn't sure myself.

mumoshu replied:
> My problem though is, that the proposal also detects changes of the Kind (e.g. Deployment -> StatefulSet).

Ah I missed this part. To me, it seems like it might benefit from having another flag to detect kind change, like `--find-kind-changes`? But not sure if that harmonizes with `--find-renames` then...

BTW, can we just omit the kind change detection? In most cases a kind change would result in a huge diff in spec and I'd say changing from a deployment to statefulset or vice versa doesn't happen so often. Rename detections sounds useful in more cases than kind change detection and complicating it for less used feature doesn't seem to pay off

fwiesel replied:
Sure, I will restrict it. That also reduces also the search-space. I just thought this is more simple to implement and covers more use-cases.

mumoshu commented on cmd/upgrade.go:
```diff
@@ -276,7 +278,7 @@ func (d *diffCmd) run() error {
 	}
 
-	seenAnyChanges := diff.Manifests(currentSpecs, newSpecs, d.suppressedKinds, d.showSecrets, d.outputContext, d.output, d.stripTrailingCR, os.Stdout)
+	seenAnyChanges := diff.Manifests(currentSpecs, newSpecs, d.suppressedKinds, d.showSecrets, d.outputContext, d.maxChangeFraction, d.output, d.stripTrailingCR, os.Stdout)
```
Can you make the rename detection get triggered only when `newInstall` is `false`? In L226-L228 we already call helm to detect if it's a diff for an upgrade(=there can be renames) or not(=there's no rename by the nature of it's a new release).

That way, I think O(N^2) issue won't affect new install badly.

fwiesel replied:
Sure, I'll change it

fwiesel commented:
When implementing the tests, I was running into some minor issues I couldn't explain, so I split up the changes into several commits and refactored the code to limit the change required to implement the functionality.
I hope, it makes it better understandable what changed and why.

mumoshu reviewed (approved):
LGTM. Thanks a lot for your efforts!
